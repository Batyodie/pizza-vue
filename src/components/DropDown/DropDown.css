@import "../../style/Base/Extend/BaseExtend.css";
.DropDown {
  position: relative;
  min-width: 231px;
}
.Label {
  @extend .BaseDropDownLabel;
}
.Btn {
  @extend .BaseDropDownLabel;
  min-width: 110px;
  text-align: left;
  border: none;
  background-color: transparent;
  cursor: pointer;
  margin: 5px;
  color: var(--accent-color);
  user-select: none;
  text-decoration: underline dotted var(--accent-color);
}

.PopUp {
  position: absolute;
  top: 30px;
  right: 0;
  max-width: 132px;
  width: 100%;
  display: flex;
  z-index: 10;
  flex-direction: column;
  background: var(--bg-content);
  box-shadow: 0px 5px 15px rgba(0, 0, 0, 0.09);
  border-radius: 10px;
  min-height: 136px;
}
.List {
  width: 100%;
  min-height: inherit;
  display: flex;
  padding-top: 12px;
  padding-bottom: 12px;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  text-align: left;

  & .Item {
    padding-left: 10px;
    padding-right: 10px;
  }
}
.Item {
  @extend .BaseDropDownLabel;
  display: flex;
  cursor: pointer;
  align-items: center;
  width: 100%;
  margin: 0;
  min-height: 38px;
  text-align: left;
  user-select: none;
  &:hover {
    background: rgba(254, 95, 30, 0.05);
  }
}

.ItemActive {
  color: var(--accent-color);
  background: rgba(254, 95, 30, 0.05);
}
