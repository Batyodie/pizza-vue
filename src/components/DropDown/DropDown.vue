<script>
import { mapGetters, mapActions } from "vuex";

import { arrowTop } from "@/assets";
import { DropDownStyle } from "@/components/style";

export default {
  name: "DropDown",

  data() {
    return {
      icon: arrowTop
    };
  },

  computed: {
    DropDownStyle() {
      return DropDownStyle;
    },

    ...mapGetters({
      getState: "getStateDropdown"
    }),

    getDropDownActive() {
      return type =>
        this.getState.DropDownItemIsActive.type === type
          ? DropDownStyle.ItemActive
          : "";
    }
  },

  mounted() {
    document.body.addEventListener("click", this.DropDownGlobalWindowClosed);
  },
  destroyed() {
    document.body.removeEventListener("click", this.DropDownGlobalWindowClosed);
  },

  methods: {
    ...mapActions({
      // proxies  `this.open()` into `this.$store.dispatch('DropDownOpened')`
      open: "DropDownOpened",
      closeGlobal: "DropDownGlobalClosed",
      selection: "DropDownSelected",
      sortPizzas: "fetchProducts"
    }),

    DropDownOpened() {
      // open DropDown
      this.open();
    },

    DropDownGlobalWindowClosed(e) {
      if (!e.path.includes(this.$refs.DropDownBlock)) {
        this.closeGlobal();
      } else {
        return null;
      }
    },

    DropDownTurned(DropDownItem) {
      // select DropDown item
      this.selection(DropDownItem);
      this.sortPizzas();
    }
  }
};
</script>

<template>
  <div ref="DropDownBlock" :class="DropDownStyle.DropDown">
    <div style="display: flex; align-items: center;">
      <img
        width="10"
        height="6"
        :src="icon"
        class="IconImage"
        alt="Arrow top"
      />

      <span :class="DropDownStyle.Label">
        <slot name="DropdownBtnLabel">Сортировка по:</slot>
      </span>
      <button @click="DropDownOpened()" :class="DropDownStyle.Btn">
        <slot name="DropdownBtnText">
          {{ getState.DropDownItemIsActive.text }}
        </slot>
      </button>
    </div>
    <div v-if="getState.DropDownIsOpen" :class="DropDownStyle.PopUp">
      <ul :class="DropDownStyle.List">
        <li
          :class="[getDropDownActive(DropDownItem.type), DropDownStyle.Item]"
          v-for="DropDownItem in getState.DropDownItems"
          :key="DropDownItem.id"
          @click="DropDownTurned(DropDownItem)"
        >
          <slot name="DropdownItemContent">
            {{ DropDownItem.text }}
          </slot>
        </li>
      </ul>
    </div>
  </div>
</template>
