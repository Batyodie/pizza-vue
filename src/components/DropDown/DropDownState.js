export default {
  state: {
    DropDownIsOpen: false,
    DropDownItemIsActive: "популярности",
    DropDownLabel: null,
    DropDownItems: [
      { id: 1, text: "популярности" },
      { id: 2, text: "по цене" },
      { id: 3, text: "по алфавиту" },
    ],
  },
  mutations: {
    DropDownOpened(state, label) {
      state.DropDownIsOpen = !state.DropDownIsOpen;
      state.DropDownLabel = label.target;
    },
    DropDownSelected(state, DropDownItem) {
      state.DropDownItemIsActive = DropDownItem.text;
      state.DropDownLabel.innerText = DropDownItem.text;
      state.DropDownIsOpen = false;
    },
    DropDownGlobalClose(state) {
      state.DropDownIsOpen = false;
    },
  },
  actions: {
    DropDownOpened(ctx, label) {
      ctx.commit("DropDownOpened", label);
    },
    DropDownGlobalClosed(ctx) {
      ctx.commit("DropDownGlobalClose");
    },
    DropDownSelected(ctx, DropDownItem) {
      ctx.commit("DropDownSelected", DropDownItem);
    },
  },
  getters: {
    getStateDropdown: (state) => {
      return state;
    },
  },
};
