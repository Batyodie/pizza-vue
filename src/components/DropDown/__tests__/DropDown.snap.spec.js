import { createLocalVue, shallowMount } from "@vue/test-utils";
import Vuex from "vuex";
// dropdwons componens files
import DropDown from "@/components/DropDown/DropDown.vue";
import DropDownState from "@/store/modules/dropdown/dropdown";

const localVue = createLocalVue();
localVue.use(Vuex);
const createStore = () => new Vuex.Store(DropDownState);

describe("General testing DropDown component", () => {
  let store = createStore();
  const wrapper = shallowMount(DropDown, {
    store,
    localVue,
  });
  const btn = wrapper.find(".Btn");

  test("Testing render dropdown component", () => {
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.html()).toBe(`<div class="DropDown">
  <div style="display: flex; align-items: center;"><img width="10" height="6" src="" alt="Arrow top"> <span class="Label">Сортировка по:</span> <button class="Btn">
      популярности
    </button></div>
  <!---->
</div>`);
  });

  test("Testing opened dropdown popUp", async () => {
    // Dropdown is closed default
    expect(store.state.DropDownIsOpen).toBe(false);

    await btn.trigger("click");
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.html()).toBe(`<div class="DropDown">
  <div style="display: flex; align-items: center;"><img width="10" height="6" src="" alt="Arrow top"> <span class="Label">Сортировка по:</span> <button class="Btn">
      популярности
    </button></div>
  <div class="PopUp">
    <ul class="List">
      <li class="ItemActive Item">
        популярности
      </li>
      <li class="Item">
        по цене
      </li>
      <li class="Item">
        по алфавиту
      </li>
    </ul>
  </div>
</div>`);
  });

  // test("Testing selected dropdown popUp item", async () => {
  //   // Open DropDown popUp
  //   store.state.DropDownIsOpen = true;
  //   expect(PopUp.isVisible()).toBe(true);
  //   expect(popUpListItem.isVisible()).toBe(true);

  //   // select all dropdown items
  //   popUpListItem.wrappers.forEach((e) => {
  //     // click dropdown item
  //     e.trigger("click");
  //     // label text in btn dropdown
  //     expect(wrapper.vm.getState.DropDownItemIsActive).toMatchSnapshot();
  //     // tests
  //     expect(store.state.DropDownItemIsActive).toContain(e.text());
  //     expect(store.state.DropDownLabel.innerText).toContain(e.text());
  //     expect(store.state.DropDownIsOpen).toBe(false);
  //   });
  // });
});
