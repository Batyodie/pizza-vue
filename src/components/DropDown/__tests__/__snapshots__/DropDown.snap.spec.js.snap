// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General testing DropDown component Testing opened dropdown popUp 1`] = `
<div
  class="DropDown"
>
  <div
    style="display: flex; align-items: center;"
  >
    <img
      alt="Arrow top"
      height="6"
      src=""
      width="10"
    />
     
    <span
      class="Label"
    >
      Сортировка по:
    </span>
     
    <button
      class="Btn"
    >
      
        популярности
      
    </button>
  </div>
   
  <div
    class="PopUp"
  >
    <ul
      class="List"
    >
      <li
        class="ItemActive Item"
      >
        
          популярности
        
      </li>
      <li
        class="Item"
      >
        
          по цене
        
      </li>
      <li
        class="Item"
      >
        
          по алфавиту
        
      </li>
    </ul>
  </div>
</div>
`;

exports[`General testing DropDown component Testing render dropdown component 1`] = `
<div
  class="DropDown"
>
  <div
    style="display: flex; align-items: center;"
  >
    <img
      alt="Arrow top"
      height="6"
      src=""
      width="10"
    />
     
    <span
      class="Label"
    >
      Сортировка по:
    </span>
     
    <button
      class="Btn"
    >
      
        популярности
      
    </button>
  </div>
   
  <!---->
</div>
`;
